src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -294,27 +296,88 @@ private static void AddIsECMABoundaryHelper(Dictionary<string, string[]> require
+                writer.WriteLine($"base.Capture(0, pos, pos);");
+            }
+            else if (root.Kind is RegexNodeKind.Nothing)
+            {
+                // Emit nothing.  This is rare in production and not something to we need optimize for, but as with
+                // empty, it's helpful as a learning exposition tool.
+                writer.WriteLine("// The pattern never matches anything.");
+            }
+            else if (root.Kind is RegexNodeKind.Multi or RegexNodeKind.One or RegexNodeKind.Notone or RegexNodeKind.Set && (root.Options & RegexOptions.IgnoreCase) == 0)
+            {
+                // If the whole expression is just one or more characters, we can rely on the FindOptimizations spitting out
+                // an IndexOf that will find the exact sequence or not, and we don't need to do additional checking beyond that.
+                needsTryFind = true;
+                using (EmitBlock(writer, "if (TryFindNextPossibleStartingPosition(inputSpan))"))
                 {
-                    writer.WriteLine("base.CheckTimeout();");
-                    writer.WriteLine();
+                    writer.WriteLine("// The search in TryFindNextPossibleStartingPosition performed the entire match.");
+                    writer.WriteLine($"int start = base.runtextpos;");
+                    writer.WriteLine($"int end = base.runtextpos = start {(!rtl ? "+" : "-")} {(root.Kind == RegexNodeKind.Multi ? root.Str!.Length : 1)};");
+                    writer.WriteLine($"base.Capture(0, start, end);");
                 }
-
-                writer.WriteLine("// If we find a match at the current position, or we have reached the end of the input, we are done.");
-                using (EmitBlock(writer, $"if (TryMatchAtCurrentPosition(inputSpan) || base.runtextpos == {(!rtl ? "inputSpan.Length" : "0")})"))
+            }
+            else if (rm.Tree.FindOptimizations.FindMode is
+                    FindNextStartingPositionMode.LeadingAnchor_LeftToRight_Beginning or
+                    FindNextStartingPositionMode.LeadingAnchor_LeftToRight_Start or
+                    FindNextStartingPositionMode.LeadingAnchor_RightToLeft_Start or
+                    FindNextStartingPositionMode.LeadingAnchor_RightToLeft_End)
+            {
+                // If the expression is anchored in such a way that there's one and only one possible position that can match,
+                // we don't need a scan loop, just a single check and match.
+                needsTryFind = needsTryMatch = true;
+                writer.WriteLine("// The pattern is anchored.  Validate the current position and try to match at it only.");
+                using (EmitBlock(writer, "if (TryFindNextPossibleStartingPosition(inputSpan) && !TryMatchAtCurrentPosition(inputSpan))"))
                 {
-                    writer.WriteLine("return;");
+                    writer.WriteLine($"base.runtextpos = {(!rtl ? "inputSpan.Length" : "0")};");
```

Comment:
To we need to do this? won't this already be set to this by TryMatchAtCurrentPosition?

Reply:
> won't this already be set to this by TryMatchAtCurrentPosition?

I don't believe so.  Where do we do that?